client/util.go

```diff
@@ -29,161 +28,76 @@ import (
 
 // TODO(pmattis): The methods in this file needs tests.
 
-func marshalKey(k interface{}) ([]byte, error) {
-	// Note that the ordering here is important. In particular, proto.Key is also
-	// a fmt.Stringer.
+func marshalKey(k interface{}) (proto.Key, error) {
 	switch t := k.(type) {
 	case string:
-		return []byte(t), nil
+		return proto.Key(t), nil
 	case proto.Key:
-		return []byte(t), nil
-	case []byte:
 		return t, nil
-	case encoding.BinaryMarshaler:
-		return t.MarshalBinary()
-	case fmt.Stringer:
-		return []byte(t.String()), nil
+	case []byte:
+		return proto.Key(t), nil
 	}
 	return nil, fmt.Errorf("unable to marshal key: %T", k)
 }
 
 // marshalValue returns a proto.Value initialized from the source
 // reflect.Value, returning an error if the types are not compatible.
-func marshalValue(v reflect.Value) (proto.Value, error) {
+func marshalValue(v interface{}) (proto.Value, error) {
 	var r proto.Value
-	if !v.IsValid() {
+	if v == nil {
 		return r, nil
 	}
 
-	switch t := v.Interface().(type) {
+	switch t := v.(type) {
 	case nil:
 		return r, nil
 
 	case string:
-		r.Bytes = []byte(t)
+		r.SetBytes([]byte(t))
 		return r, nil
 
 	case []byte:
-		r.Bytes = t
+		r.SetBytes(t)
 		return r, nil
 
 	case proto.Key:
-		r.Bytes = []byte(t)
+		r.SetBytes([]byte(t))
 		return r, nil
 
-	case gogoproto.Message:
-		var err error
-		r.Bytes, err = gogoproto.Marshal(t)
+	case time.Time:
+		err := r.SetTime(t)
 		return r, err
 
-	case encoding.BinaryMarshaler:
-		var err error
-		r.Bytes, err = t.MarshalBinary()
+	case gogoproto.Message:
+		err := r.SetProto(t)
 		return r, err
 	}
 
-	switch v.Kind() {
+	switch v := reflect.ValueOf(v); v.Kind() {
 	case reflect.Bool:
```

Comment:
Yes. If you pass `parser.DBool` to one of the `client` routines, it would be great for it to marshal correctly.